providers/kubeconfig/provider.go

```diff
@@ -0,0 +1,593 @@
+		opts.KubeconfigPath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	}
+
+	return &Provider{
+		opts:        opts,
+		log:         log.Log.WithName("kubeconfig-provider"),
+		client:      nil, // Will be set in Run
+		clusters:    map[string]cluster.Cluster{},
+		cancelFns:   map[string]context.CancelFunc{},
+		seenHashes:  map[string]string{},
+		readySignal: make(chan struct{}),
+	}
+}
+
+// Options are the options for the Kubeconfig Provider.
+type Options struct {
+	// Namespace to watch for kubeconfig secrets
+	Namespace string
+
+	// Label key to identify kubeconfig secrets
+	KubeconfigLabel string
+
+	// Key in the secret data that contains the kubeconfig
+	KubeconfigKey string
+
+	// ConnectionTimeout is the timeout for connecting to a cluster
+	ConnectionTimeout time.Duration
+
+	// CacheSyncTimeout is the timeout for waiting for the cache to sync
+	CacheSyncTimeout time.Duration
+
+	// KubeconfigPath is the path to the kubeconfig file to use for development/testing
+	// If not set, will use the default ~/.kube/config
+	KubeconfigPath string
+}
+
+type index struct {
+	object       client.Object
+	field        string
+	extractValue client.IndexerFunc
+}
+
+// Provider is a cluster provider that watches for secrets containing kubeconfig data
+// and engages clusters based on those kubeconfigs.
+type Provider struct {
+	opts        Options
+	log         logr.Logger
+	client      client.Client
+	lock        sync.RWMutex
+	clusters    map[string]cluster.Cluster
+	cancelFns   map[string]context.CancelFunc
+	indexers    []index
+	seenHashes  map[string]string // tracks resource versions
+	readySignal chan struct{}     // Signal when provider is ready to start
+	readyOnce   sync.Once         // Ensure we only signal once
+}
+
+// IsReady returns a channel that will be closed when the provider is ready to start
+func (p *Provider) IsReady() <-chan struct{} {
+	return p.readySignal
+}
+
+// Get returns the cluster with the given name, if it is known.
+func (p *Provider) Get(ctx context.Context, clusterName string) (cluster.Cluster, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if cl, ok := p.clusters[clusterName]; ok {
+		return cl, nil
+	}
+
+	return nil, fmt.Errorf("cluster %s not found", clusterName)
+}
+
+// Run starts the provider and blocks, watching for kubeconfig secrets.
+func (p *Provider) Run(ctx context.Context, mgr mcmanager.Manager) error {
+	p.log.Info("Starting kubeconfig provider", "namespace", p.opts.Namespace, "label", p.opts.KubeconfigLabel)
+
+	// If client isn't set yet, get it from the manager
+	if p.client == nil && mgr != nil {
+		p.log.Info("Setting client from manager")
+		p.client = mgr.GetLocalManager().GetClient()
+		if p.client == nil {
+			p.log.Error(nil, "Failed to get client from manager, will use direct API access")
+		}
+	}
+
+	// Set up Kubernetes config for API operations - we'll use this for direct API access
+	// This bypasses the controller-runtime cache which might not be ready yet
```

Comment:
Why not call `cache.WaitForCacheSync` to make sure the cache is ready?

Reply:
Done